sources/aws-smithy-experimental/src/hyper_1_0.rs

```diff
@@ -209,6 +213,60 @@ mod build_connector {
     ) -> hyper_rustls::HttpsConnector<HttpConnector<HyperUtilResolver<R>>> {
         make_tls(HyperUtilResolver { resolver }, crypto_provider.provider())
     }
+
+    pub(super) fn https_with_proxy(
+        https_connector: hyper_rustls::HttpsConnector<HttpConnector>,
+        https_proxy: &str,
+        no_proxy: Option<Vec<String>>,
+    ) -> hyper_http_proxy::ProxyConnector<hyper_rustls::HttpsConnector<HttpConnector>> {
+        // Determines whether a request of a given scheme, host and port should be proxied
+        // according to `https_proxy` and `no_proxy`.
+
+        let intercept = move |scheme: Option<&str>, host: Option<&str>, _port| {
+            if let Some(host) = host {
+                if let Some(no_proxy) = &no_proxy {
+                    if scheme != Some("https") {
+                        return false;
+                    }
+                    if no_proxy.iter().any(|s| s == "*") {
+                        // Don't proxy anything
+                        return false;
+                    }
+                    // If the host matches one of the no proxy list entries, return false (don't proxy)
+                    // Note that we're not doing anything fancy here for checking `no_proxy` since
+                    // we only expect requests here to be going out to some AWS API endpoint.
+                    return !no_proxy.iter().any(|no_proxy_host| {
+                        !no_proxy_host.is_empty() && host.ends_with(no_proxy_host)
+                    });
+                }
+                true
+            } else {
+                false
+            }
+        };
+
+        let mut proxy_uri = https_proxy.parse::<Uri>().expect("Invalid proxy URI");
+
+        // If the proxy's URI doesn't have a scheme, assume HTTP for the scheme and let the proxy
+        // server forward HTTPS connections and start a tunnel.
+        if proxy_uri.scheme().is_none() {
+            proxy_uri = format!("http://{}", https_proxy)
+                .parse::<Uri>()
+                .expect("Unable to parse proxy URI as HTTPS");
+        }
+        let mut proxy = Proxy::new(intercept, proxy_uri);
+        // Parse https_proxy as URL to extract out auth information if any
+        let proxy_url = Url::parse(https_proxy).expect("Unable to parse HTTPS proxy as URL");
+
+        if !proxy_url.username().is_empty() || proxy_url.password().is_some() {
```

Comment:
nit: I was thinking this might read more clearly if it were
```suggestion
        if !(proxy_url.username().is_empty() && proxy_url.password.is_none()) {
```

since the only case where we don't set auth is when both username and password aren't provided

Reply:
This seems fine, but since this has been working as is for a bit I'd like to keep this particular PR to just the necessities.